### ISSUE TITLE
cFE Application Developers Guide Should be Doxygen/Markdown Based

### ISSUE THREAD

**Issue by skliper** [nasa/cFE#202]
As discussed at the 5-24-16 CCB meeting (reference cFS Systems Engineering Handbook discussion in the minutes), the cFE application developers guide contains a significant amount of API references.  Indirectly referencing the cFS API's in an external document is prone to error.  The API information should be referenced directly from the code.  It is highly recommended to move the Application Developers Guide to the cFE Doxygen User’s Guide.

**Comment by skliper** [comment]
_Imported from trac issue 171.  Created by sstrege on 2016-07-08T12:27:02, last modified: 2019-07-03T12:48:08_

**Comment by skliper** [comment]
_Trac comment by sstrege on 2016-09-15 20:50:25:_

This guide should also include information on how to write endian neutral code:

On Thursday, September 15, 2016 7:10:10 PM EDT Scott, Michael S. (ARC-TI)[SGT, INC] wrote:
> In our project we used this:
> 
> INCLUDE(TestBigEndian)
> 
> TEST_BIG_ENDIAN(BIGENDIAN)
> IF(${BIGENDIAN})
>      ADD_DEFINITIONS(-DBIGENDIAN)
> ENDIF(${BIGENDIAN})
> 

This is the correct way to do it in cmake, but as Joe said there isn't really much reason to know this in cmake. 

With regard to Joe's email, a "better" way to do this in C is:

#define IS_LITTLE_ENDIAN (((union {unsigned x; char c;}){1}).c)

This is known at compile time, which is nice. So you can even do something like

#if IS_LITTLE_ENDIAN
...
#else
...
#endif

Steven

**Comment by skliper** [comment]
_Trac comment by glimes on 2016-11-08 14:17:08:_

Current crop of cfe-next are all going into CFE 6.6

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-07-03 12:48:08:_

Moved unfinished 6.6.1 issues to next minor release

**Comment by skliper** [comment]
Note #361 added a markdown version, but conversion to doxygen would make it linkable and reduce duplicated documentation.

**Comment by jphickey** [comment]
> Note #361 added a markdown version, but conversion to doxygen would make it linkable and reduce duplicated documentation.

Can't the markdown file itself just be included in the doxygen?  Doxygen supports using markdown files and including that in the generated HTML, and terms should be cross-linked as expected.

**Comment by skliper** [comment]
Absolutely, although I'd suggest another scrub in that context to remove duplicated documentation.
